### PR TITLE
feat: Add composite indexes

### LINKED ISSUE #299: Sec. Indexes: Implement composite indexes
Part of https://github.com/sourcenetwork/defradb/issues/201

### PR DESCRIPTION
**Pull request by islamaliev**
## Relevant issue(s)

Resolves #299 

## Description

This change introduces composite secondary indexes as well as unique-composite secondary indexes.
Note: different order direction for composite fields is not included in this change.

## Tasks

- [x] I made sure the code is well commented, particularly hard-to-understand areas.
- [x] I made sure the repository-held documentation is changed accordingly.
- [x] I made sure the pull request title adheres to the conventional commit style (the subset used in the project can be found in [tools/configs/chglog/config.yml](tools/configs/chglog/config.yml)).
- [x] I made sure to discuss its limitations such as threats to validity, vulnerability to mistake and misuse, robustness to invalidation of assumptions, resource requirements, ...

## How has this been tested?

Unit and integration tests.

Specify the platform(s) on which this was tested:
- MacOS

### CODE REVIEW

File: db/fetcher/indexer_iterators.go
Comment (AndrewSisley) [inline db/fetcher/indexer_iterators.go:187]: question: I don't see a constructor for `eqSingleIndexIterator`, where is it guaranteed that the slice will have a `0` index?

File: db/fetcher/indexer_iterators.go
Comment (AndrewSisley) [inline db/fetcher/indexer_iterators.go:405]: praise: Thanks for tidying this switch up - looks much nicer :)

File: db/fetcher/indexer_iterators.go
Comment (AndrewSisley) [inline db/fetcher/indexer_iterators.go:407]: question: As mentioned, I really do like the improvements to this switch, however it does feel a bit strange that you are returning an interface and instead of having lots of implementations for that interface, you are passing in a func to a private property in a concrete implementation of that interface and then returning that one 'super' type instance.

Why did you chose to do this?  Did it just kind of evolve as you were working in this space?

File: db/fetcher/indexer_iterators.go
Comment (AndrewSisley) [inline db/fetcher/indexer_iterators.go:441]: todo: Please declare this error in the errors.go file :)

File: db/fetcher/indexer_iterators.go
Comment (AndrewSisley) [inline db/fetcher/indexer_iterators.go:469]: todo: I can't spot any good reason for this ever happening, and codecov says it is untested.  Should this not be an error?  Or perhaps can we cast here without checking type?  If neither of those can you please add a comment documenting how this can happen, and why it is okay if it does.

EDIT: When looking at the next `fieldInd != propKey.Index` if, I think this is just you cycling through trying to find the correct property, however the cast raises eyebrows and it's intent is not that clear - if this the reason for this code, can I suggest that instead of the current you do the below, as it makes it clearer that the cast is just a requirement for finding the correct prop:

```
if !ok || fieldInd != propKey.Index {
    // this is not the property we are looking for (reword :))
    continue
}
```

**Review by AndrewSisley — COMMENTED** [review]
Part way through my review, submitting comments so far :)

File: db/fetcher/indexer_iterators.go
Comment (AndrewSisley) [inline db/fetcher/indexer_iterators.go:577]: todo: This looks like it will always be false no? As to get here at least one of the `fieldContitions` must be `opIn`, and `isUniqueFetchByFullKey` requires that all `fieldContitions` must be `opEq`.

Regardless of whether I'm right or wrong here, can you please document this.

File: db/index.go
Comment (AndrewSisley) [inline db/index.go:99]: suggestion:  I don't think the new error kind is worth it here, or at least it should wrap our `client.ErrFieldNotExist` error.  The more different kinds of errors we have in the system, the harder it will be for users to know/discover/handle them. Happy for this to be shrugged off as out of scope, as this was here before.

File: db/indexed_docs_test.go
Comment (AndrewSisley) [inline db/indexed_docs_test.go:97]: nitpick: I think `fields' names` implies that each field has multiple names.  If each field only has one name then `field names` is more appropriate.  Same goes for the variable names below.

File: planner/filter/normalize.go
Comment (AndrewSisley) [inline planner/filter/normalize.go:188]: question: Why has this changed?  What else could it be?

File: planner/select.go
Comment (AndrewSisley) [inline planner/select.go:300]: suggestion: Inverting this if statement would make it easier to read IMO.

File: planner/select.go
Comment (AndrewSisley) [inline planner/select.go:299]: suggestion:  Is an old name I know, but it looks like it is missing a word? I suggest renaming to `findFilteredIndexByIndexedField`, or maybe `findIndexByFilteringField`.

File: planner/select.go
Comment (AndrewSisley) [inline planner/select.go:309]: question:  This looks suspicious, why are you just returning the first?  What if there are multiple indexes using the same field? It that tested?

**Review by AndrewSisley — COMMENTED** [review]
Nearly through it, submitting a few more comments.

File: tests/integration/index/query_with_composite_index_only_filter_test.go
Comment (AndrewSisley) [inline tests/integration/index/query_with_composite_index_only_filter_test.go:56]: suggestion: I do feel like you are hiding the most important part of this test here.  The index is what you are testing, and as reader, what I am most interesting in seeing, and yet it is not visible within the test.

File: tests/integration/index/query_with_composite_index_only_filter_test.go
Comment (AndrewSisley) [inline tests/integration/index/query_with_composite_index_only_filter_test.go:429]: suggestion: I dislike how you had to pull the other super important part of the test out of the action flow, disrupting things for the reader.

I wonder if you could add an action that wraps the two requests perhaps, so the test could read:
```
testUtils.RequestWithExplain{
    Request: `query {
		User(filter: {age: {_nin: [20, 23, 28, 42]}, name: {_nin: ["John", "Andy", "Chris"]}}) {
			name
		}
	}`,
	Results: []map[string]any{
					{"name": "Roy"},
					{"name": "Islam"},
					{"name": "Keenan"},
				},
	Explanation: testUtils.NewExplainAsserter().WithFieldFetches(3).WithIndexFetches(10),
}
``` 
Internally it would split the action into the two Request Actions (that are currently here in the test) before sending them all into the main test executor func.

Happy if you prefer we dont do this in this PR.

File: tests/integration/test_case.go
Comment (AndrewSisley) [inline tests/integration/test_case.go:250]: suggestion: It feels a bit clunky to have to separate fields from directions like this.  Happy for it to be done outside of this PR but I do think it would be nicer to have something like the below:
```
Fields []client.IndexedFieldDescription
```

File: db/fetcher/indexer_iterators.go
Comment (islamaliev) [reply db/fetcher/indexer_iterators.go:42]: it's just there for composite indexes. We construct a slice of value matchers with every matcher being responsible for a corresponding field in the index to match. For some fields there might not be any criteria to match. For examples if you have composite index of /name/age/email/ and in the filter you specify only "name" and "email". Then the "_any" matcher will be used for "age".

I will add a comment to explain this.

Comment (AndrewSisley) [reply db/fetcher/indexer_iterators.go:42]: Okay cheers, I think I'm understanding this better.  In that case, can I also suggestion:

suggestion: Can those that are actual request filters reference those consts, and those which are not (i.e. opAny) remain as strings - this might make it clearer that some are special (and will also may it easier and less error prone to change the public filter syntax as these values will update when those do).  e.g.:
```
const (
   opEq = request.FilterOpEq,
   ...
   opAny   = "_any"
)
```
Although I found that `connor`/etc doesn't have a const yet for these, which would mean you have the hassle of changing all of those too, so feel very free to shrug this suggestion off as out of scope.

File: client/index.go
Comment (islamaliev) [reply client/index.go:62]: added a comment, renamed

File: db/errors.go
Comment (islamaliev) [reply db/errors.go:90]: merged 2 errors into 1

File: db/fetcher/indexer_iterators.go
Comment (islamaliev) [reply db/fetcher/indexer_iterators.go:187]: it is created in `createIndexIterator` and it's guaranteed because it's the field value we are seeking and the value is being  set upon creation.

File: db/fetcher/indexer_iterators.go
Comment (islamaliev) [reply db/fetcher/indexer_iterators.go:407]: the interface is different kinds of matchers there is byte-matcher, in-matcher, like-matcher and new any-matcher. All of them (except for any) are configurable. Byte-matcher just a bit more: it compares bytes and we can configure what is the expected result of the comparison. 

I think this way it's a bit more concise, but I don't have a strong preference here.

File: db/fetcher/indexer_iterators.go
Comment (islamaliev) [reply db/fetcher/indexer_iterators.go:441]: done

File: db/fetcher/indexer_iterators.go
Comment (AndrewSisley) [reply db/fetcher/indexer_iterators.go:407]: Ah okay, there are still a fair few concrete types then :)

> I think this way it's a bit more concise, but I don't have a strong preference here.

Keep it as-is then, unless anyone else has a really strong preference - I have nothing really against this, it just struck me as unusual and worth highlighting/chatting about :)  Is a fairly nice way of reducing the number to types required.

File: db/fetcher/indexer_iterators.go
Comment (islamaliev) [reply db/fetcher/indexer_iterators.go:469]: merged ifs and added a comment

File: db/fetcher/indexer_iterators.go
Comment (islamaliev) [reply db/fetcher/indexer_iterators.go:577]: added a comment to `isUniqueFetchByFullKey`

File: db/index.go
Comment (islamaliev) [reply db/index.go:99]: replaced the error

File: db/indexed_docs_test.go
Comment (islamaliev) [reply db/indexed_docs_test.go:97]: adjusted

File: planner/filter/normalize.go
Comment (islamaliev) [reply planner/filter/normalize.go:188]: `normalizeProperty` method that calls this `normalizeProperties` checks is given condition is `map[connor.FilterKey]any` or an array of conditions which has a type of `[]any`. But as it recursively checks all nested conditions and values one of the values (not conditions) can match `[]any` (for example "_in" op). That's why we need to check if it's really a conditions or just value.

File: planner/select.go
Comment (islamaliev) [reply planner/select.go:300]: adjusted. Not part of this PR though

File: planner/select.go
Comment (islamaliev) [reply planner/select.go:309]: it will be part of another PR. 

Added a comment.

File: tests/integration/index/query_with_composite_index_only_filter_test.go
Comment (islamaliev) [reply tests/integration/index/query_with_composite_index_only_filter_test.go]: done

File: tests/integration/index/query_with_composite_index_only_filter_test.go
Comment (islamaliev) [reply tests/integration/index/query_with_composite_index_only_filter_test.go:475]: we've got a task for it https://github.com/sourcenetwork/defradb/issues/1962

File: tests/integration/test_case.go
Comment (islamaliev) [reply tests/integration/test_case.go:250]: I don't think I agree. This will make every such test action more wordy. There will be a few tests that test different directions.

I just added a comment there. You were not supposed to comment on this part :D 

The next task I'm working on involves different directions. I'd suggest we check how it's going to look in those tests and discuss it there.

File: planner/filter/normalize.go
Comment (AndrewSisley) [reply planner/filter/normalize.go:188]: Ah okay - this was a bug fix?  Did you add an integration test that covers this?

File: planner/select.go
Comment (AndrewSisley) [reply planner/select.go]: cheers Islam

File: planner/select.go
Comment (AndrewSisley) [reply planner/select.go]: Okay cheers.  Can you please make sure that there are integration tests that make sure that if there are multiple overlapping indexes (e.g. two composites, a simple and a composite) that queries still work (no need for explain test documenting this issue, but that would be nice if you can do it reliably without much effort too)?

File: tests/integration/index/query_with_composite_index_only_filter_test.go
Comment (AndrewSisley) [reply tests/integration/index/query_with_composite_index_only_filter_test.go:475]: Ah nice, my bad and thanks :)

File: tests/integration/test_case.go
Comment (AndrewSisley) [reply tests/integration/test_case.go:250]: > I just added a comment there. You were not supposed to comment on this part :D

:grin: :grin: 

> The next task I'm working on involves different directions. I'd suggest we check how it's going to look in those tests and discuss it there.

That sounds sensible, I'm happy with that :)

File: db/fetcher/indexer_iterators.go
Comment (AndrewSisley) [reply db/fetcher/indexer_iterators.go:192]: Ah okay, I would prefer a constructor for stuff like that, but `createIndexIterator` is a factory-like func and  the concrete a private type, so I am happy for this to be shrugged off as my personal preference here :grin:

File: db/fetcher/indexer_iterators.go
Comment (AndrewSisley) [reply db/fetcher/indexer_iterators.go]: cheers, looks good

File: db/fetcher/indexer_iterators.go
Comment (AndrewSisley) [reply db/fetcher/indexer_iterators.go:587]: Ah of course, cheers - that comment really helps, I didn't spot it skipping the first index :)

**Review by AndrewSisley — APPROVED** [review]
Looks good to me, thanks Islam :)

I'm approving now, as all but one of my remaining comments are very optional within this PR. I do think we need the following though before merge please:

> Can you please make sure that there are integration tests that make sure that if there are multiple overlapping indexes (e.g. two composites, a simple and a composite) that queries still work (no need for explain test documenting this issue, but that would be nice if you can do it reliably without much effort too).

And preferably at least one for what looks like a bug fix in `filter/normailze.go` (if it is a bug fix :))

File: planner/filter/normalize.go
Comment (islamaliev) [reply planner/filter/normalize.go:188]: an integration test found this bug :)

File: planner/select.go
Comment (islamaliev) [reply planner/select.go]: This line of code is not new. There are integration tests that cover it

**Pull request merged**
